src/vision/status_quo/alan_runs_into_stack_trouble.md

```diff
@@ -0,0 +1,60 @@
+
+Looking back at the documents made it clear to Alan that this could potentially be the issue: `pin_mut!` pins futures to the stack. It was relatively clear that a possible solution would be to pin to the heap instead of the stack. Some more digging in the docs lead Alan to `Box::pin` which did just that. An extra heap allocation was of no consequence to him, so he gave it a try. Lo and behold, this fixed the issue! 
+
+While Alan knew enough about pinning to know how to satisfy the compiler, he didn't originally take the time to fully understand what the consequences were of using `pin_mut!` to pin his futures. Now he knows!
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * When coming from a background of GCed languages, taking the time to understand the allocation profile of a particular piece of code is not something Alan was used to doing.
+    * It was hard to tell where in his code the stack was being exhausted. Alan had to rely on manually combing his code to find the culprit.
+    * Pinning is relatively confusing, and although the code compiled, Alan didn't fully understand what he wrote and what consequences his decision to use `pin_mut!` would have.
+* **What are the sources for this story?**
+    * This story is adapted from the experiences of the team working on the [Krustlet](https://github.com/deislabs/krustlet) project. You can read about this story in their own words [here](https://deislabs.io/posts/a-heaping-helping-of-stacks/).
+* **Why did you choose Alan to tell this story?**
+    * The programmers this story was based on have an experience mostly in Go, a GCed language.
+    * The story is rooted in the explicit choice of using stack vs heap allocation, a choice that in GCed languages is not in the hands of the programmer.
+* **How would this story have played out differently for the other characters?**
+    * Grace would have likely had a similar hard time with this bug. While she's used to the tradeoffs of stack vs heap allocations, the analogy to the `Pin` API is not present in languages she's used to.
+    * Barbara, as an expert in Rust, may have had the tools to understand that `pin_mut` is used for pinning to the stack while `Box::pin` is for pinning heap allocations.
+    * This problem is somewhat subtle, so someone like Niklaus would probably have had a much harder time figuring this out (or even getting the code to compile in the first place).
+* **Could Alan have used another API to achieve the same objectives?**
+    * Perhaps! Tokio's `select!` macro doesn't require explicit pinning of the futures it's provided, but it's unclear to this author whether it would have been smart enough to avoid pinning large futures to the stack. However, pinning is a part of the way one uses futures in Rust, so it's possible that such an issue would have arisen elsewhere.
```

Comment:
It just pins things on the "async stack" automatically for you. The futures-rs version however does that too after I implemented https://github.com/rust-lang/futures-rs/pull/1873 . The tokio version mainly goes a little bit further and avoids the `FusedFuture` requirements based off the ideas in https://github.com/rust-lang/futures-rs/issues/1989.

I don't any version of the `select!` macro would resolve the issue on its own - one will always need an explicit `Box` to decouple the `Future`s. Only something like `FuturesUnordered` or other sub-executors will do it automatically.

Still surprised that the whole thing took 4Gb and could be fixed by improving one select. [While it's known that generators can grow exponentially in size](https://github.com/rust-lang/rust/issues/69826) it seems to take a while to get to 4Gb 😯